src/rust/engine/options/src/config.rs

```diff
@@ -217,21 +217,132 @@ fn toml_table_to_dict(table: &Value) -> HashMap<String, Val> {
 }
 
 #[derive(Clone)]
-pub(crate) struct Config {
+struct ConfigSource {
+    #[allow(dead_code)]
+    path: Option<OsString>,
     config: Value,
 }
 
+impl ConfigSource {
```

Comment:
The diff looks scarier than it is. The meat of this change is the introduction of a ConfigSource, representing a single Config, and moving the generic get_value/get_list/get_dict logic to it unchanged. Then Config has a vec of these ConfigSources, and does the right thing.

The rest is just test and plumbing changes.